### DISCUSSION TITLE
docs: Best practice using Dtos with Entity constructors

### DISCUSSION THREAD

**Discussion by conspireagency** [mikro-orm/mikro-orm#3835] [General]
I was wondering what the best practice is when combining Dtos and Entity constructors.  Say you're taking some input from the frontend into the backend and save it to the database.

Is something like below possible?
`
const upsell = new Upsell(createUpsellDto);
await this.em.persistAndFlush(upsell);
`
Or would you need to do something like this:

`
const variants = createUpsellDto.variants.map((variant) =>
      this.em.getReference(Variant, variant),
    );

    const upsell = new Upsell(
      this.em.getReference(Shop, createUpsellDto.shop),
      <Displays>createUpsellDto.displays,
      <Trigger>createUpsellDto.trigger,
      createUpsellDto.triggerIds,
      this.em.getReference(Product, createUpsellDto.product),
      variants,
    );

    await this.em.persistAndFlush(upsell);
`

Below is an example of the entity with what I was thinking for the constructor, but doesn't work (expected)

`@Entity()
export class Upsell {
    
  @PrimaryKey()
  id!: number;

  @ManyToOne(() => Shop)
  shop!: Shop;

  // this is the area in which the upsell displays
  @Enum(() => Displays)
  displays!: Displays;

  // this is the area in which the upsell displays
  @Enum(() => Trigger)
  trigger!: Trigger;

  // this can hold product or collection ids depending on the conditions of the trigger
  @Property({ nullable: true })
  triggerIds: string[];

  // this is the product that is shown in the upsell
  @ManyToOne(() => Product)
  product!: Product;

  // this is the product variants is shown in the upsell
  @ManyToOne(() => Variant)
  variants: Variant[];

  constructor(
    createupsellDto: CreateUpsellDto,
  ) {
    this.shop = createupsellDto.shop;
    this.displays = createupsellDto.displays;
    this.trigger = createupsellDto.trigger;
    this.triggerIds = createupsellDto.triggerIds;
    this.product = createupsellDto.product;
    this.variants = createupsellDto.variants;
  }
}

export enum Displays {
  ATC = 'atc',
  CART = 'cart',
  BLOG = 'blog',
}

export enum Trigger {
  ALL = 'all',
  PRODUCT = 'specific_product',
  COLLECTION = 'specific_collection',
}`

**Comment by B4nan** [comment]
This is purely a matter of preference, some like to keep empty constructors and use `em.create()`, and some like to have a constructor listing the required params, so like to have a single param constructor with a DTO. All of it is possible and should work mostly the same. Note that `em.create()` will understand all of those variants and extract the constructor parameters automatically.

If your example compiles but doesn't work, you are doing something wrong (e.g. lying to type system) probably. Not sure what is the `CreateUpsellDto` definition, but I guess you assign raw PKs or POJOs to relation entity properties (given your `em.getReference` example) - that's not possible. `em.create()` is where all of this happens automatically, but that's not really about constructors.

My go-to way would be to use actual entity properties in your constructor, so you don't have this problem on the entity level, but above, where it's usually easily solvable. But in your case, if you just want to build entity graph from a DTO, `em.create()` is the way to go, regardless of your constructor I guess.

Note there is also `em.assign()` which handles the same, and `Reference.createNakedFromPK()` (will need to integrate that one to `ref()` helper probably). The `createNakedFromPK` is quite a handy method, as that is what allows you to convert entity PK to entity instance without knowing the context (= entity manager). In other words, the things that were previously covered only via `em.create` and `em.assign()` could be done inside the entity constructor. I will think about making this possible in v6.

There are also many users trying to use the `assign` helper from inside the constructor to handle this (same as `em.assign`, but works without EM reference unless you try to work with relations - and that is exactly the case that could be now solved without the EM reference, via `createNakedFromPK`). 

So generally speaking, with `em.create` you dont have to care much about your constructors, and if your plan is to feed them with DTOs, its better to go with `em.create` and keep constructors empty. The conversion from POJOs to entity instances needs to happen, and it won't happen automatically, you can't just assign POJOs to the entity properties directly.

**Comment by B4nan** [comment]
Made some improvements that will be part of v5.6, and added a good portion of examples around this here: https://github.com/mikro-orm/mikro-orm/pull/3837

**Reply by conspireagency** [reply]
@B4nan this is awesome, you're the man!
